### ISSUE TITLE
feature request - increase work area in browser

### ISSUE THREAD

**Issue by lklepner** [nypl-spacetime/nypl-warper#11]
Working on a large screen I've found the map working areas frustratingly small in comparison to the available space in the browser -
![screen shot 2016-03-03 at 12 09 52 pm](https://cloud.githubusercontent.com/assets/110027/13502419/e0b8aa2e-e138-11e5-8aa8-3b20dabdc87a.png)

As a quick fix on my end I create a chrome add-on which maximizes the screen area. I put a gist up to share the code, which is pretty simple at this point, in case there is some way to work this into the project itself -
https://gist.github.com/lklepner/f7a6fd689fc6798adce2 

Here are some screenshots of the end result -
![screen shot 2016-03-03 at 12 16 38 pm](https://cloud.githubusercontent.com/assets/110027/13502763/4e3bc756-e13a-11e5-9634-8a661a6a4fd2.png)

![screen shot 2016-03-03 at 12 19 49 pm](https://cloud.githubusercontent.com/assets/110027/13502768/56d53622-e13a-11e5-9bcf-a600e0402d17.png)

![screen shot 2016-03-03 at 12 18 15 pm](https://cloud.githubusercontent.com/assets/110027/13502772/5fc58520-e13a-11e5-8c34-5b6e5d542127.png)

**Comment by mgiraldo** [comment]
awesome... we agree that the current window size is very limited...

thanks

**Comment by lklepner** [comment]
If there was an interest I would be happy to prepare a proper pull request

**Comment by mgiraldo** [comment]
that would be great!

**Comment by lklepner** [comment]
I've cloned the project and run vagrant up. Is there a trick to getting the
port forwarding to work for the HTTP requests from the host?

I'm able to connect to ssh through the port forwarding but haven't been
able to connect to http://127.0.0.1:3000 from my host.  If I curl directly
from the guest I see html in the response so I believe the server is
running properly..

Any ideas?

Thanks! I look forward to work on this together.

**Comment by mgiraldo** [comment]
if you `curl 127.0.0.1:3000` all is well but not if you open it in the browser? what response do you get in curl? what do you get in browser? a blank page?

**Comment by lklepner** [comment]
Starting rails with -b 0.0.0.0 did the trick. I'll submit a PR with the
changes.

**Comment by lklepner** [comment]
PR submitted. I'm still running into another problem - nothing from the
assets folder is loading. Everything except the page itself is coming back
404.

**Comment by lklepner** [comment]
Seems like the assets 404s might be tied into the hostname / ip problem I
ran into when initially starting up -

Started GET "/warper/assets/jquery.nouislider.all.min.js?body=1" for
10.0.2.2 at 2016-03-07 18:56:51 +0000

Cannot render console from 10.0.2.2! Allowed networks: 127.0.0.1, ::1,
127.0.0.0/127.255.255.255

**Comment by lklepner** [comment]
The initial problem had been an empty response at port 3000. Adding -b
0.0.0.0 caused the main page to load but the assets are still coming back
404.

It looks like the rails server config might be throwing an error here as
well -
Cannot render console from 10.0.2.2! Allowed networks: 127.0.0.1, ::1,
127.0.0.0/127.255.255.255

**Comment by lklepner** [comment]
Adding a whitelist for 10.0.2.2 to the development.rb config file cleared
up the render errors -

  class Application < Rails::Application
    config.web_console.whitelisted_ips = '10.0.2.2'
  end

I'm not sure if you want this change in a PR.

Initial map data - is there a way for me to obtain a dump of map data to
restore locally?

**Comment by mgiraldo** [comment]
the 404s might be because it is looking in the `/warper` folder. i tried running without the prefix and changing this in `config/environments/development.rb`:

```
config.action_controller.relative_url_root = ""
config.action_mailer.default_url_options = { :script_name => "" }
```

running the server now loads the CSS properly

**Comment by lklepner** [comment]
Thanks for the tips on the 404s. I think i've got everything running now.
Next up I need to figure out how to load sample data.. and then a sample
user account (i see notes in the readme) so i can access the interface as
an authenticated user.

**Comment by mgiraldo** [comment]
the readme does not tell you to add a `confirmed_at` date. do so or you will be unable to login

(many things to improve in the docs ¯\_(ツ)_/¯)

i guess to load sample data is best to upload some raster map for now.

**Comment by lklepner** [comment]
I'll try creating an account and uploading a raster map. I'd also be happy
to help clean up the docs as I go through things.

I'm experiencing extremely slow response times when moving page to page on
the local dev instance of the project - often times pages take a minute or
longer to load all of their assets.  Is there something I can do to address
this?

**Comment by mgiraldo** [comment]
the thing is the Vagrant version is virtualized so it basically creates a mini Ubuntu and runs it. if you feel brave you can try installing all dependencies manually in your computer (it's a lot of dependencies). if you know Vagrant you can try giving more resources to that instance (i don't know Vagrant).

docs help would be great

**Comment by lklepner** [comment]
I was able to speed up vagrant by increasing the memory and switching to
NFS for the fileshare.

The next stumbling block is getting the DB initialized -

vagrant@vagrant-ubuntu-trusty-64:~$ sudo su - postgres
postgres@vagrant-ubuntu-trusty-64:~$ psql mapwarper_dev -c "create
extension postgis;"
psql: FATAL:  database "mapwarper_dev" does not exist

Its been ages since I've done anything with postgres and even then I wasn't
working in the console. Can you advise?

**Comment by mgiraldo** [comment]
if you do vagrant you dont need to do any of that... it "just works" and you only do:

```
cd /srv/mapwarper
rails c
```
create the user and exit
```
rails s
```

**Comment by lklepner** [comment]
Gotcha - so at the rails c irb prompt I should paste in the db
initialization below?

user = User.new
user.login = "super"
user.email = "super@superxyz123.com"
user.password = "your_password"
user.password_confirmation = "your_password"
user.activated_at = Time.now
user.confirmed_at = Time.now
user.save

role = Role.find_by_name('super user')
user = User.find_by_login('super')

permission  = Permission.new
permission.role = role
permission.user = user
permission.save

role = Role.find_by_name('administrator')
permission = Permission.new
permission.role = role
permission.user = user
permission.save

**Comment by mgiraldo** [comment]
correct

that creates a user with login `super@superxyz123.com` and pass `your_password`

**Comment by lklepner** [comment]
Excellent - i'm in now! Many thanks for walking me through this. It wasn't
clear to me from the readme that the db initialization should take place
after the rails c command.

Can you advise on how to load a raster map to get the ball rolling? is
there any docs on this?

**Comment by mgiraldo** [comment]
this codebase only supports maps from nypl. fortunately, many of our maps are public domain so you can test it. for this you need a Digital Collections API account and token from http://api.repo.nypl.org/. the warper imports (subject to throttling limits) based on the map UUID visible in the [item page](http://digitalcollections.nypl.org/items/a66456f6-3860-95d5-e040-e00a18065ecc) (`a66456f6-3860-95d5-e040-e00a18065ecc` in this case):

1. add the generated api token to the `/config/nypl_repo.yml` file ([example file](https://github.com/nypl-spacetime/nypl-warper/blob/master/config/nypl_repo.yml.example))
2. re-start the server and log in with a super user
3. go to Admin / Imports
4. Add New Import
5. Select Import Type: Map and type in the UUID: `a66456f6-3860-95d5-e040-e00a18065ecc` and click Create
6. Click Start Import

this will take a few minutes to pull the map. you now have a map you can work with and test code.

**Comment by lklepner** [comment]
Thanks for the outline. I've got the map imported. Would you like me to add
these notes to the readme file?

**Comment by mgiraldo** [comment]
i'm adding these notes right now... about to re-test and push

**Comment by mgiraldo** [comment]
[done](https://github.com/nypl-spacetime/nypl-warper#development)

**Comment by lklepner** [comment]
Do you have any preference where the calculation for the expanded layout
would be added into the mix?

I've been leaning towards /app/views/layouts/mapdetail.html.erb lines 49-83

**Comment by mgiraldo** [comment]
that makes sense

**Comment by lklepner** [comment]
I'm getting an error when refreshing the maps page after applying a crop
outline -
Error in loading GML file
http://mapwarper.livefeed.net:3000/shared/masks/1.gml.ol?1459536456

You can try it here -
http://mapwarper.livefeed.net:3000/warper/maps/1#Crop_tab

Any idea about the cause/resolution here?

**Comment by mgiraldo** [comment]
url not opening for me

**Comment by mgiraldo** [comment]
this may have to do with the `src_maps_dir` and `dst_maps_dir` which by default are in the `/public/mapimages` folder. if you customized that folder in `config/application.yml` then you need to make sure it is writable and all that

**Comment by lklepner** [comment]
appplication.yml has default values at this point.

I see files are being created in the mapimages folder

public/mapimages/1.gfs
public/mapimages/1.gml
public/mapimages/1.gml.ol
public/mapimages/dst/1.tif
public/mapimages/dst/png/1.png
public/mapimages/dst/png/1.png.aux.xml
public/mapimages/src/a66456f6-3860-95d5-e040-e00a18065ecc.tif
public/mapimages/src/a66456f6-3860-95d5-e040-e00a18065ecc.tif_masked

**Comment by mgiraldo** [comment]
🤔

is the gml valid? did you crop it in the crop tab? i still cannot view the URL

**Comment by lklepner** [comment]
The .gml and .gml.ol files look valid at first glance -

.gml
<wfs:FeatureCollection
xmlns:wfs='http://www.opengis.net/wfs'><gml:featureMember
xmlns:gml='http://www.opengis.net/gml'><feature:features xmlns:feature='
http://mapserver.gis.umn.edu/mapserver'
fid='OpenLayers.Feature.Vector_2156'><feature:geometry><gml:Polygon><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates
cs=',' decimal='.' ts=' '>1488.548,496.17199999999957
3132.084,898.1880000000001 2753.716,3369.404 1269.804,3712.3
1098.356,3339.844 1263.892,1991.908 1293.4520000000023,1421.400000000002
1399.8680000000004,706.0479999999961
1488.548,496.17199999999957</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></feature:geometry></feature:features></gml:featureMember></wfs:FeatureCollection>

.gml.ol

<wfs:FeatureCollection
xmlns:wfs="http://www.opengis.net/wfs"><gml:featureMember
xmlns:gml="http://www.opengis.net/gml"><feature:features xmlns:feature="
http://mapserver.gis.umn.edu/mapserver"
fid="OpenLayers.Feature.Vector_2156"><feature:geometry><gml:Polygon><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates
decimal="." cs="," ts=" ">1488.548,4915.828 3132.084,4513.812
2753.716,2042.596 1269.804,1699.7 1098.356,2072.156 1263.892,3420.092
1293.4520000000023,3990.599999999998 1399.8680000000004,4705.952000000004
1488.548,4915.828</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></feature:geometry></feature:features></gml:featureMember></wfs:FeatureCollection>

Are you able to log the page itself?

**Comment by mgiraldo** [comment]
i get:
```
Unable to connect
Firefox can't establish a connection to the server at mapwarper.livefeed.net:3000.
```
in Firefox and:
```
This site can’t be reached
mapwarper.livefeed.net refused to connect.
```
in Chrome

removing the port number doesn't help either

**Comment by lklepner** [comment]
Hmmm. Its a new setup but I tested it from a remote computer and was able
to load http://mapwarper.livefeed.net:3000/warper/maps/1

**Comment by mgiraldo** [comment]
seems to be a NYPL network issue. i can access it through a VPN… VPN is a bit slow/glitchy so I cannot test very well… will try back home

**Comment by lklepner** [comment]
Any luck from home? I was hoping to use this server to review frontend
changes together.

**Comment by mgiraldo** [comment]
i created an account but had an error... maybe you can go to the admin and approve me as a user?

**Comment by lklepner** [comment]
account approved and set to role super.

**Comment by mgiraldo** [comment]
i drew a mask and started the masking process. but i still need a vpn to view this url and my vpn is not very stable so it is hard to know what is going on. does it work properly in localhost? could it be a config problem in livefeed?

**Comment by lklepner** [comment]
I'm able to load things on my local network and over VNC using a client's
remote computer. I'm guessing the NYPL outbound firewall is blocking
requests to port 3000? I can try moving things over to port 80. I only have
a single IP and have a few things running on it so I'll need to setup a
reverse proxy through the other VM.

**Comment by mgiraldo** [comment]
ok i think i found it:

the `/shared/masks` folder is hardcoded so it is not taking into account the real location

🤔

if the config file points to some network drive (which is the case for the current production server), a mapping needs to be done to point that path to the drive

if it is just the default (empty) url then no need to use `/shared/masks` but `/warper/mapimages/1.gml.ol`

i added a new [branch](https://github.com/nypl-spacetime/nypl-warper/tree/crop) with a quick fix to this, although still hardcoded to `/warper/` but would work for instances that use that extra folder structure

**Comment by lklepner** [comment]
That did the trick - thanks!

**Comment by mgiraldo** [comment]
one thing we could do is increase the default height (maybe twice as much as it is right now) and disable the scroll-to-zoom feature in the map… it is annoying and we already have a zoom in/out button/slider

this would make it a very simple fix

**Comment by lklepner** [comment]
This sounds like a good starting point. The scroll to zoom is extremely
erratic as currently implemented. Ive been reading the openlayers docs to
see if there is a way to slow it down..

**Comment by mgiraldo** [comment]
dont bother slowing it down... straight out disable would be ideal...

**Comment by lklepner** [comment]
Sounds good. I'll give it a try in the morning.

**Comment by lklepner** [comment]
I've made some modifications to the Show tab:
- removed zoomwheel (by disabling MouseDefaults control for this map)
- moved resize functionality into unwarped.js so its instantiated when the
tab is first opened
- setup resize listener to adjust layout when browser is resized
- adjusted the maxResolution so the unwarped_image fills the available space

http://mapwarper.livefeed.net/warper/maps/1

The other tabs haven't been changed at this point other than changing the
default height from 500px to 1000px.

How does this look on your end?

**Comment by mgiraldo** [comment]
seems to be down

**Comment by lklepner** [comment]
My router is on its last legs.. I gave it a reboot and it seems to be
working now. Want to give it another whirl?

**Comment by mgiraldo** [comment]
looks great

**Comment by lklepner** [comment]
I expanded the layout changes to the rectify and crop tabs (zoomwheel
change is not yet in)

- animating transitions between showing the complete header for the Show
tab vs maximizing the workspace for Rectify and Crop tabs
- improving the scroll position calculation basing it on the wooTabs
position rather than an arbitrary number which became inaccurate if the map
title wrapped to multiple lines

http://mapwarper.livefeed.net/warper/maps/1

If this is heading in a good direction from your standpoint I can continue
with more tabs.

Question on insets - is there a way to name them?

**Comment by mgiraldo** [comment]
looking good! can you make the animation a bit quicker?

insets inherit their parent metadata so no, can't be renamed :\

**Comment by lklepner** [comment]
Thanks for the positive feedback. I've sped up the animations from 2000ms
to 500ms. I'll start working on the rest of the tabs.

Insets are pretty confusing without the ability to assign a name to them -
they look like duplicates of the original map in the map listings / align
tool. Would it make sense to create a new issue / feature request to
address this at some point?

**Comment by lklepner** [comment]
Another quick question - can you point me towards some documentation on how
the align tool works?

I've been hesitant to try it since I'm unsure how the overwrite and type of
alignment options will play out.

Am I correct in expecting the alignment tool will enable the original map
and multiple insets to be viewed as a single map?

**Comment by mgiraldo** [comment]
re insets: yes... make a new issue about it... or maybe the issue is that the “insetness” of an inset is not readily apparent... if there were a better way of signaling this... in the UI... maybe: `metadata title (inset 1)`

re alignment tool: we don't need to do the fancy resizing there. i think what it does is allow for maps that are close to each other in the same `layer` to use their overlapping control points. see [this layer](http://maps.nypl.org/warper/layers/859) for example

**Comment by lklepner** [comment]
If I understand correctly.. the align tool isn't really applicable to the
combining multiple insets. Are insets completely independent maps other
than the metadata and image?

I've completed the Crop and Rectify tabs, disabling the zoomwheel and
cleaning up the associated JS. I think we're in pretty good shape at this
point so I've submitted a PR.

https://github.com/nypl-spacetime/nypl-warper/pull/21
